File: archive/f/f-sharp/JobSequencing.fs
open System

[<Struct>]
type Job = { Profit: int; Deadline: int }


module JobSequencing =
    let rec private findAvailable (parent: int array) (x: int) =
        if parent.[x] <> x then
            parent.[x] <- findAvailable parent parent.[x]

        parent.[x]

    let solve (profits, deadlines) =
        let jobs =
            List.map2 (fun p d -> { Profit = p; Deadline = d }) profits deadlines
            |> List.sortByDescending (fun j -> j.Profit)

        if List.isEmpty jobs then
            0
        else
            let maxDeadline = jobs |> List.maxBy (fun j -> j.Deadline) |> _.Deadline
            let slots = Array.init (maxDeadline + 1) id

            jobs
            |> List.fold
                (fun total job ->
                    let slot = findAvailable slots (min job.Deadline maxDeadline)

                    if slot > 0 then
                        slots.[slot] <- findAvailable slots (slot - 1)
                        total + job.Profit
                    else
                        total)
                0

module Helpers =
    let usage = "Usage: please provide a list of profits and a list of deadlines"

    let private (|IntList|_|) (s: string) =
        s.Split(',', StringSplitOptions.RemoveEmptyEntries)
        |> Array.choose (fun p ->
            match p.Trim() |> Int32.TryParse with
            | true, n -> Some n
            | _ -> None)
        |> fun arr -> if arr.Length > 0 then Some(Array.toList arr) else None

    let parseArgs argv =
        match argv with
        | [| IntList p; IntList d |] when p.Length = d.Length -> Ok(p, d)
        | _ -> Error usage

    let handleResults =
        function
        | Ok result ->
            printfn "%d" result
            0
        | Error msg ->
            eprintfn "%s" msg
            1


[<EntryPoint>]
let main argv =
    argv
    |> Helpers.parseArgs
    |> Result.map JobSequencing.solve
    |> Helpers.handleResults

File: archive/v/visual-basic/job-sequencing.vb
Public Class Job
    Public Property Profit As Integer
    Public Property Deadline As Integer

    Public Sub New(profit As Integer, deadline As Integer)
        Me.Profit = profit
        Me.Deadline = deadline
    End Sub
End Class


Public Module JobSequencing

    Public Sub Main(args As String())

        If args.Length <> 2 Then
            Console.WriteLine("Usage: please provide a list of profits and a list of deadlines")
            Return
        End If

        Dim profits = ParseIntList(args(0))
        Dim deadlines = ParseIntList(args(1))

        If profits Is Nothing OrElse deadlines Is Nothing OrElse
           profits.Count <> deadlines.Count Then

            Console.WriteLine("Usage: please provide a list of profits and a list of deadlines")
            Return
        End If

        Dim jobs As New List(Of Job)

        For i As Integer = 0 To profits.Count - 1
            jobs.Add(New Job(profits(i), deadlines(i)))
        Next

        Dim selected = GetMaxProfitJobSequence(jobs)

        Dim totalProfit As Integer = 0

        For Each job In selected
            totalProfit += job.Profit
        Next

        Console.WriteLine(totalProfit)

    End Sub


    Private Function ParseIntList(input As String) As List(Of Integer)

        If String.IsNullOrWhiteSpace(input) Then
            Return Nothing
        End If

        Dim result As New List(Of Integer)

        For Each part In input.Split(","c, StringSplitOptions.RemoveEmptyEntries)

            Dim value As Integer

            If Not Integer.TryParse(part.Trim(), value) Then
                Return Nothing
            End If

            result.Add(value)

        Next

        Return result

    End Function


    Public Function GetMaxProfitJobSequence(jobs As List(Of Job)) As List(Of Job)

        jobs.Sort(Function(a, b) b.Profit.CompareTo(a.Profit))

        Dim maxDeadline As Integer = 0

        For Each job In jobs
            If job.Deadline > maxDeadline Then
                maxDeadline = job.Deadline
            End If
        Next

        Dim timeSlots(maxDeadline - 1) As Boolean
        Dim selected As New List(Of Job)

        For Each job In jobs

            For i As Integer = job.Deadline - 1 To 0 Step -1

                If Not timeSlots(i) Then
                    timeSlots(i) = True
                    selected.Add(job)
                    Exit For
                End If

            Next

        Next

        Return selected

    End Function

End Module